src/Palette.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2023 Inclusive Design Research Centre, OCAD University
+ * All rights reserved.
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/inclusive-design/adaptive-palette/blob/main/LICENSE
+ */
+
+import { render } from "preact";
+import { html } from "htm/preact";
+import { PaletteCell } from "./PaletteCell";
+
+/**
+ * Given a palette defined in a json structure, compute the number of rows
+ * and columns in that palette.
+ *
+ * @param {Object} paletteDefinition - An object that lists the positions,
+ *                 heights and widths of the cells in the palette.
+ * @return {Object} - The row and column counts: `{ numRows: ..., numColumns: ...}`.
+ */
+function countRowsColumns (paletteDefinition) {
+  let rowCount = 0;
+  let colCount = 0;
+  let rightColumn = 0;
+  let bottomRow = 0;
+  const cellIds = Object.keys(paletteDefinition.cells);
+  cellIds.forEach((id) => {
+    const aCell = paletteDefinition.cells[id];
+    rightColumn = aCell.columnStart + aCell.columnSpan;
+    if (rightColumn > colCount) {
+      colCount = rightColumn;
+    }
+    bottomRow = aCell.rowStart + aCell.rowSpan;
+    if (bottomRow > rowCount) {
+      rowCount = bottomRow;
+    }
+  });
+  return { numRows: rowCount, numColumns: colCount };
+}
+
+export function Palette (props) {
```

Comment:
A typescript `type` can be defined to verify the data structure of `props`.

The same comment applies to other component props.

Reply:
I was going to write an interface for the `props` arguments, and give them a better name, but I was putting that off until the structure was better defined.  So, I will define that interface at some point.